### PR TITLE
Improve Docker Compose container execution

### PR DESCRIPTION
**Pull request by endorama**
This PR has been extracted from https://github.com/elastic/elastic-package/pull/665.

When `elastic-package` runs docker containers it relies on [`WaitForHealthy()`](https://github.com/endorama/elastic-package/blob/89161065b946c1531132f0e20c03b4f48210f0b7/internal/compose/compose.go#L267) to ensure expected containers are running.

This functions has a missing feature: timing out if the condition is not met.

This PR adds this feature, adding a 10 minute timeout after which an error is returned.

To prevent waiting time waiting for a condition that can never be met a check is added for unrecoverable termination condition (container is reported as exited and with an exit code greater than 0) to immediately exit with an error, skipping waiting until timeout expiration.

### CODE REVIEW

File: internal/compose/compose.go
Comment (endorama) [inline internal/compose/compose.go:316]: The same line have been added in https://github.com/elastic/elastic-package/pull/688/commits/8563a1c8a7fec6375d82520d423ce33abf331b8f#diff-8019656c843609e3cb15ca57ab96dc41fdadae8215d0ff22d9b989f795aef4a4 so this PR will need to be rebased on top of `main` once https://github.com/elastic/elastic-package/pull/688 is merged

Comment (endorama) [reply internal/compose/compose.go:317]: removed

File: internal/compose/compose.go
Comment (jsoriano) [inline internal/compose/compose.go:279]:
```suggestion
	ticker := time.NewTicker(1 * time.Second)
	defer ticker.Stop()
```

File: internal/compose/compose.go
Comment (jsoriano) [inline internal/compose/compose.go:323]: :+1:

File: internal/compose/compose.go
Comment (jsoriano) [inline internal/compose/compose.go:278]: I was going to suggest to use `context.WithTimeout()`, so we can also interrupt calls to docker if they hang, but leveraging this context would require more changes in `internal/docker`.

File: internal/compose/compose.go
Comment (jsoriano) [inline internal/compose/compose.go:291]: Nit. Do nothing and continue in this `case`, to have the main part of the loop in the upper level.
```suggestion
		case <-ticker.C:
		}
		healthy := true
```

File: internal/compose/compose.go
Comment (jsoriano) [inline internal/compose/compose.go:316]: :+1:

**Review by jsoriano — COMMENTED** [review]
Ticker needs to be stopped, for the rest it LGTM :+1:

File: internal/compose/compose.go
Comment (mtojek) [inline internal/compose/compose.go:278]: Please move this timeout to const at the top of the file and name it accordingly.

File: internal/compose/compose.go
Comment (mtojek) [inline internal/compose/compose.go:289]: nit: timeout waiting...

File: internal/compose/compose.go
Comment (mtojek) [inline internal/compose/compose.go:287]: Let me challenge this and consider another approach.

We already have "for" loop with a 1-sec delay per iteration. The function can store `time.Now()` and just compare it if the `startTime.Add(timeout)` is greater than `time.Now`. We don't need to add channels/select/ticker. WDYT?

Comment (endorama) [reply internal/compose/compose.go:287]: I spent some time searching differences between `time.Sleep` and `time.Tick`, and the only thing I found is that a tick tries to guarantee the interval, while sleep is unaware of execution times. There seems to be no difference in performance for the 2 implementations.

In general I prefer the ticker approach as it's behaviour is explicit, but in this case I don't see particular benefits as guaranteeing the interval is not a concern.

I'll rework it with your approach.

Comment (jsoriano) [reply internal/compose/compose.go:287]: I think I prefer the ticker and select approach too. It also works better than time arithmetic when you have multiple source of events, as may happen if/when we have contexts here.

Comment (jsoriano) [reply internal/compose/compose.go:287]: But we can wait to have contexts here, not a strong opinion :slightly_smiling_face:

Comment (mtojek) [reply internal/compose/compose.go:287]: I'm a bit confused here as adding a for-select block will introduce another nesting level and make the function hard to read. TBH I would follow the KISS rule, so if it's possible to solve with a single if-clause, I would go for it.

In general, we can consider refactoring of this area, as there is a similar logic in the [SIGINT](https://github.com/elastic/elastic-package/blob/8d9732d0c4364d28ba1c34f2ad7fe91dfc7d2785/internal/signal/sigint.go#L24) function. It would be great, If we can refactor all parts (SIGINT, timeout, waiter logic, contexts). You have my sword :)

Comment (endorama) [reply internal/compose/compose.go:287]: By running a quick search I found there are 5 `for {` loops in the codebase. I think it would make a lot of sense to generalize the approach, so interval/timeout/signals are handled consistently. 

Are you ok to keep it for a future PR?  I already have a ticker-based `Waiter` struct implementation to use as a baseline, but I'm blocked on other work streams so I would prefer to push this refactor to the back burner for a bit.

File: internal/compose/compose.go
Comment (endorama) [reply internal/compose/compose.go]: the bug is obvious, I misinterpreted the logic. `healthy` must be reinitialized at each loop, otherwise once it become `false` the logic does not change it's value anymore. I tried to use it to detect the timeout, but I stumbled upon the bug. The current implementation is clearer and uses times instead of a side effect on the `healthy` value.

File: internal/compose/compose.go
Comment (mtojek) [inline internal/compose/compose.go:291]: nit: let's leave this empty line

File: internal/compose/compose.go
Comment (mtojek) [inline internal/compose/compose.go:30]: nit: I think you can make those consistent:

```
const (
   waitForHealthyTimeout = ...
   waitForHealthyInterval = ...
)
```

File: internal/compose/compose.go
Comment (jsoriano) [inline internal/compose/compose.go:27]: Nit. No need to convert to `time.Duration`, a number multiplying a duration is already a duration.

File: internal/compose/compose.go
Comment (jsoriano) [inline internal/compose/compose.go:288]: Nit. Break the loop directly with a return so we don't have to remember to check for the timeout again after the loop? (And we are consistent on how we handle other end conditions as `SIGINT`).
```suggestion
	for {
		if time.Now().After(timeout) {
			return errors.New("timeout waiting for healthy container")
		}
```

**Review by jsoriano — COMMENTED** [review]
I don't have anything against `for {`s I am ok with leaving them if they make sense :slightly_smiling_face:

**Review by mtojek — APPROVED** [review]
LGTM

**Review by jsoriano — APPROVED** [review]

**Pull request merged**
